numpy/f2py/tests/test_array_from_pyobj.py

```diff
@@ -304,10 +305,16 @@ def test_in_out(self):
         assert_(not intent.in_.is_intent('c'))
 
 
-class _test_shared_memory(object):
+class TestSharedMemory(object):
     num2seq = [1, 2]
     num23seq = [[1, 2, 3], [4, 5, 6]]
 
+    @pytest.fixture(autouse=True, scope='class', params=_type_names)
+    def setup_type(self, request):
+        request.cls.type = Type(request.param)
+        request.cls.array = lambda self, dims, intent, obj: \
+            Array(Type(request.param), dims, intent, obj)
+
```

Comment:
We do not encourage the use of fixtures, but I think in this case it might make sense, although I find I have to stare at both the old and the new implementations for quite a while to understand what they are doing. Maybe a high-level comment that explains what is going on here and how this generates multiple tests would help? Like (and I probably am going to mess this up)

Generate a set of TestSharedMemory classes, each with a different `type` and `array` attribute derved from `_type_names`. (and then how does `params=type_names` become `request.param`? Is the kwarg  `param` magic? What happened to the `s`?. Isn't `autouse` True by default, why is it needed here?)

Reply:
A slightly different way to write this is:
```python
      @pytest.fixture(autouse=True, params=_type_names)
      def setup_type(self, request):
          self.type = Type(request.param)
          self.array = lambda dims, intent, obj: \
              Array(Type(request.param), dims, intent, obj)
```
which is a bit closer to the way the `eval`'d version looked. The only difference is it gets called for each method instead of once per class. I can change it to this if it's clearer (+/- a comment.)

Converting `params=_type_names` is how [fixtures are parametrized](https://docs.pytest.org/en/latest/fixture.html#fixture-parametrize). The `request` parameter is [described here](https://docs.pytest.org/en/latest/reference.html#request) (side note: I really hate that the pytest docs don't properly cross-reference the objects.) There's no `s` because it's called for each parameter. `autouse` is False by default (the default idiom is to explicitly name used fixtures as parameters), so it must be specified to get it applied to the whole class.

Reply:
I think the class setup is better, if there is more than one class-method it will be faster.

Maybe this could be expanded upon in our [testing](https://github.com/numpy/numpy/blob/master/doc/TESTS.rst.txt#easier-setup-and-teardown-functions-methods) recommendations in a different PR, with an annotated example. 

@charris, any thoughts? Does replacing the `eval` justify a fixture?

Reply:
Given that we will probably be using pytest for many years, I don't think we should go out of our way to avoid fixtures if they are the cleanest option. I do like setup/teardown where appropriate because of clarity, but that may just be familiarity on my part. The f2py tests are pretty old, so if they can be made better with modern tools, that is a good thing. That said, I haven't looked at this PR yet :)

Reply:
@mattip  Try updating pytest.

Reply:
Works here. Might also need to clean out the build directory if you do a lot of testing.

Reply:
Now working. 

> what are the generated test names for your proposal

The test names all get `[<dtype>]` tacked to the end of the name, so they become (broken into two pieces)
```
numpy/f2py/tests/test_array_from_pyobj.py::TestSharedMemory:: \
test_f_in_from_23casttype[CLONGDOUBLE]
```